DEP: remove inheritance to `QMCEngine` in `MultinomialQMC` and `MultivariateNormalQMC`

tupui opened the pull request:
Follow up of #15865

* Remove inheritance from `QMCEngine` with QMC dimensions: `MultinomialQMC` and `MultivariateNormalQMC`
* Fix bug with dimension when `inv_transform=False`

This is a breaking change, but arguably nobody would notice it since the feature is very new and I could not see any use.

tupui commented:
Wait the multivariate does not sound good after more thinking. I will rework this.

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -799,7 +799,12 @@ def integers(
             sample = self.random(n=n)
 
-        sample = scale(sample, l_bounds=l_bounds, u_bounds=u_bounds)
+        # scaling, no check of hypercube as sample are generated by engine
```
Can this go back to using `scale` now that `MultinomialQMC` and  `MultivariateNormalQMC` won't use this?

mdhaber replied:
@tupui?

tupui replied:
Forgot! will fix this now.

mdhaber reviewed (approved):
This is fine modulo one comment, but please take a look at https://github.com/tupui/scipy/pull/5 as an alternative approach.

tupui commented:
@mdhaber I updated the PR with the following:

* Removed `fast_forward` and `reset` from both QMC distributions
* Updated the tests
* Fixed a bug about dimension checking when `inv_transform=False`. I added tests.

I don't like the current logic of the tests. When I get some time, I will see in another PR to streamline that.

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -1680,9 +1684,11 @@ def __init__(
-                                 " with dimensions of mean and covariance.")
+                                 " with dimensions of mean and covariance."
+                                 " If `inv_transform` is False, it must be"
+                                 " equal to `2 * ceil(d / 2)`")
```
```suggestion
                                 " an even number.")
```

mdhaber commented on scipy/stats/tests/test_qmc.py:
```diff
@@ -830,40 +830,6 @@ def test_other_engine(self):
-                                               engine=base_engine,
-                                               inv_transform=True)
-            samples = engine.random()
-            assert_equal(samples.shape, (1, d))
```
Is this test still reasonable to keep?

mdhaber commented on scipy/stats/tests/test_qmc.py:
```diff
@@ -923,6 +881,18 @@ def test_NormalQMCSeeded(self):
+        # same test with another engine
+        seed = np.random.default_rng(274600237797326520096085022671371676017)
+        base_engine = qmc.Sobol(4, scramble=True, seed=seed)
+        engine = qmc.MultivariateNormalQMC(
```
At some point, variable names should be changed to distributions are not referred to as engines.

tupui replied:
Good point.

mdhaber commented on scipy/stats/tests/test_qmc.py:
```diff
@@ -923,6 +881,18 @@ def test_NormalQMCSeeded(self):
                                      [-0.997875, 0.399134, 1.032234]])
         assert_allclose(samples, samples_expected, atol=1e-4)
 
+        # same test with another engine
```
Is it really another engine, or is an equivalent engine that happens to be passed in directly rather than it being created internaly?

tupui replied:
It's the same engine so that we have the same output. I used the same wording as with the other tests for now.

I tried to not do much on the test and just did the minimum to reflect the changes, but this is definitely not in a good state and should be reworked.

mdhaber commented:
OK, other than the above, I'd suggest checking with the mailing list to see if they agree with the idea of changing this without a deprecation cycle.

tupui commented:
Mail sent: https://mail.python.org/archives/list/scipy-dev@python.org/thread/LMOGIMMF6DLSB7TP6JU75IXV4Q5P6Q3N/

h-vetinari commented:
> This is a breaking change, but arguably nobody would notice it since the feature is very new and I could not see any use.

Has there been any released version with this? https://github.com/scipy/scipy/pull/15865 was only added for 1.9.0, so I don't think that's even a concern? If something did break relative to a release version, then I think we should at least mention it in the release notes.

tupui commented:
> > This is a breaking change, but arguably nobody would notice it since the feature is very new and I could not see any use.
> 
> 
> 
> Has there been any released version with this? https://github.com/scipy/scipy/pull/15865 was only added for 1.9.0, so I don't think that's even a concern? If something did break relative to a released version, then I think we should at least mention it in the release notes.

The inheritance is there since 1.7—when we initially released QMC. But we argue that here it does not need a deprecation cycle as these classes are not widely used and the capabilities it removes is not something you would do as an end user.

In any case there would be a mention in the release notes.

tupui commented:
Friendly reminder @mdhaber
